modules/calib3d/src/chessboard.cpp

```diff
@@ -0,0 +1,3116 @@
+ // (including, but not limited to, procurement of substitute goods or services;
+ // loss of use, data, or profits; or business interruption) however caused
+ // and on any theory of liability, whether in contract, strict liability,
+ // or tort (including negligence or otherwise) arising in any way out of
+ // the use of this software, even if advised of the possibility of such damage.
+ //
+ //M*/
+
+#include "precomp.hpp"
+// #include "opencv2/imgproc.hpp"
+#include "opencv2/flann.hpp"
+#include "chessboard.hpp"
+
+using namespace std;
+namespace cv
+{
+    /////////////////////////////////////////////////////////////////////////////
+    /////////////////////////////////////////////////////////////////////////////
+    // magic numbers used for chessboard corner detection
+    /////////////////////////////////////////////////////////////////////////////
+    const float CORNERS_SEARCH = 0.5;                 // percentage of the edge length to the next corner used to find new corners
+    const float MAX_ANGLE = 45.0/180.0*M_PI;          // max angle between line segments supposed to be straight
+    const float MIN_COS_ANGLE = cos(35.0/180*M_PI);   // min cos angle between board edges
+    const float MIN_RESPONSE_RATIO = 0.3;
+    const float ELLIPSE_WIDTH = 0.35;                 // width of the search ellipse in percentage of its lenght
+    const float RAD2DEG = 180.0/M_PI;
+    /////////////////////////////////////////////////////////////////////////////
+    /////////////////////////////////////////////////////////////////////////////
+
+    // some helper methods
+    bool isPointOnLine(cv::Point2f l1,cv::Point2f l2,cv::Point2f pt,float min_angle);
+    int testPointSymmetry(cv::Mat mat,cv::Point2f pt,float dist,float max_error);
+    double getEdgePos(cv::InputArray _data);
+    double calcSubpixel(const double &x_l,const double &x,const double &x_r);
+    double calcSubPos(const double &x_l,const double &x,const double &x_r);
+    void polyfit(const Mat& src_x, const Mat& src_y, Mat& dst, int order);
+    float calcSignedDistance(const cv::Vec2f &n,const cv::Point2f &a,const cv::Point2f &pt);
+    void normalizePoints1D(cv::InputArray _points,cv::OutputArray _T,cv::OutputArray _new_points);
+    cv::Mat findHomography1D(cv::InputArray _src,cv::InputArray _dst);
+
+    // some helper methods
+//    cv::Point2f operator/(const cv::Point2f &a, const float &b)
+//    {
+//        return cv::Point2f(a.x/b,a.y/b);
+//    }
+//    cv::Point2f operator*(const cv::Point2f &a, const float &b)
+//    {
+//        return cv::Point2f(a.x*b,a.y*b);
+//    }
+//
+    void normalizePoints1D(cv::InputArray _points,cv::OutputArray _T,cv::OutputArray _new_points)
+    {
+        cv::Mat points = _points.getMat();
+        if(points.cols > 1 && points.rows == 1)
+            points = points.reshape(1,points.cols);
+        if(points.channels() != 1 )
+            throw std::runtime_error("findHomography: only data with one channel are supported");
```

Comment:
It is better to use `CV_Error()` macro to raise errors. Or even one from `CV_Check()` macro